### PR TITLE
types.py: add TPM2B_OBJECT to handle size + buffer structures

### PR DESCRIPTION
**Pull request by whooo**
* Makes the size attribute read only
* Keeps track of buffer size when buffer is changed
* Makes buffer data inmutable
* Makes the class indexable, iterable and sliceable

I had to change unrelated tests as they where using TPM2B_DIGEST

### CODE REVIEW

**Comment by whooo** [comment]
I'll rebase on master if everything looks good to you

**Comment by whooo** [comment]
ok, I now see there is already an commit for this...

**Comment by whooo** [comment]
@williamcroberts, so this to a large degree overwrites your work, sorry about that, I was almost done with this version.
I skipped all TPM2B structures which aren't a size + a byte array, as size seems ignored by the API for those (I never set the correct size for publicArea in TPM2B_PUBLIC for example).

I wanted to make the byte array immutable, but memoryview.toreadonly() requires python 3.8 or later.
It also supports the structures that have another field then buffer (TPM2B_NAME for example)

There is also a small change to the API, _cdata doesn't come first now, so it needs to be specified (see the change in ESAPY.py), as I think it's useful to be able to to TPM2B_DIGEST(bunch_of_bytes)

**Comment by williamcroberts** [comment]
> @williamcroberts, so this to a large degree overwrites your work, sorry about that, I was almost done with this version.
> I skipped all TPM2B structures which aren't a size + a byte array, as size seems ignored by the API for those (I never set the correct size for publicArea in TPM2B_PUBLIC for example).

Yeah the general paradigm is that callers intentionally leave that to zero, and then libmu figures out the right size.

> 
> I wanted to make the byte array immutable, but memoryview.toreadonly() requires python 3.8 or later.
> It also supports the structures that have another field then buffer (TPM2B_NAME for example)

I thought the current tip of master does that, I had it look for count/size and then the other field is the one you care about.

> 
> There is also a small change to the API, _cdata doesn't come first now, so it needs to be specified (see the change in ESAPY.py), as I think it's useful to be able to to TPM2B_DIGEST(bunch_of_bytes)

I just figured it out by type of data

File: tpm2_pytss/types.py [hunk 1]
```diff
@@ -1460,15 +1463,15 @@ class TPM2B_ATTEST(TPM2B_OBJECT):
     pass
 
 
-class TPM2B_CONTEXT_DATA(TPM2B_OBJECT):
+class TPM2B_CONTEXT_DATA(TPM_OBJECT):
```
Comment (williamcroberts) [inline tpm2_pytss/types.py:1466]: Why are we no longer inheriting? These are TPM2B structures.

Comment (whooo) [reply tpm2_pytss/types.py]: Seems I looked at the wrong structure, fixed it now.

**Comment by whooo** [comment]
> > @williamcroberts, so this to a large degree overwrites your work, sorry about that, I was almost done with this version.
> > I skipped all TPM2B structures which aren't a size + a byte array, as size seems ignored by the API for those (I never set the correct size for publicArea in TPM2B_PUBLIC for example).
> 
> Yeah the general paradigm is that callers intentionally leave that to zero, and then libmu figures out the right size.
> 
> > I wanted to make the byte array immutable, but memoryview.toreadonly() requires python 3.8 or later.
> > It also supports the structures that have another field then buffer (TPM2B_NAME for example)
> 
> I thought the current tip of master does that, I had it look for count/size and then the other field is the one you care about.
I was thinking about that only "buffer" as key is being checked in `__getattribute__` and checking for TPM2B_DIGEST ini __init__
> 
> > There is also a small change to the API, _cdata doesn't come first now, so it needs to be specified (see the change in ESAPY.py), as I think it's useful to be able to to TPM2B_DIGEST(bunch_of_bytes)
> 
> I just figured it out by type of data

I see that now, will revert to that part

File: tpm2_pytss/types.py [hunk 2]
```diff
@@ -1516,7 +1517,7 @@ class TPM2B_NAME(TPM2B_OBJECT):
     pass
 
 
-class TPM2B_NV_PUBLIC(TPM2B_OBJECT):
+class TPM2B_NV_PUBLIC(TPM_OBJECT):
```
Comment (williamcroberts) [inline tpm2_pytss/types.py:1520]: I guess it's simpler to avoid the complex TPM2B's then handling it in the TPM2B_OBJECT class itself. Can we rename TPM2B_OBJECT to TPM2B_SIMPLE_OBJECT to better clarify that only simple, non compound TPM2B's should inherit from it.

File: test/test_types.py [hunk 3]
```diff
@@ -584,8 +580,7 @@ def test_unmarshal(self):
         d, offset = TPM2B_DIGEST.Unmarshal(buf)
         self.assertEqual(offset, 7)
         self.assertEqual(d.size, 5)
-        db = d.buffer
-        self.assertEqual(db, b"test1")
+        self.assertEqual(d.buffer, b"test1")
```
Comment (williamcroberts) [inline test/test_types.py:583]: Make this cleanup a separate commit.

**Review by williamcroberts — CHANGES_REQUESTED** [review]
This looks a lot better, just some small changes.

File: tpm2_pytss/types.py [hunk 4]
```diff
@@ -921,54 +921,55 @@ def Unmarshal(cls, buf):
...
 
     def __bytes__(self):
-
-        _cdata = self._cdata
-
-        # not everything uses .buffer, for instance TPM2B_NAME uses .name
-        tipe = cpointer_to_ctype(_cdata)
-        field_name = next((v[0] for v in tipe.fields if v[0] != "size"), None)
-
-        buffer = TPM2B_unpack(_cdata, n=field_name)
```
Comment (williamcroberts) [inline tpm2_pytss/types.py:964]: TPM2B_unpack is no longer used remove from imports

File: tpm2_pytss/types.py [hunk 5]
```diff
@@ -921,54 +921,55 @@ def Unmarshal(cls, buf):
...
-
-        # update the len
-        _cdata.size = len(value)
+    def __setattr__(self, key, value):
+        _bytefield = type(self)._get_bytefield()
+        if key == "size":
+            raise AttributeError(f"{key} is read only")
+        super().__setattr__(key, value)
+        if key == _bytefield:
+            self._cdata.size = len(value)
```
Comment (williamcroberts) [inline tpm2_pytss/types.py:941]: LGTM now complains in the __init__ for TPM_OBJECT that:
`NEWCall to self.__setattr__ in __init__ method, which is overridden by method TPM2B_OBJECT.__setattr__.`

We should probably have the caller in TPM_OBJECT call:
_cdata.__setattr_(k, v) to avoid that.

**Review by williamcroberts — CHANGES_REQUESTED** [review]
A few things from LGTM. Id like to pull this in as it blocks some ESYS testing, any chance you can make these changes soon, or can I just edit your commit?

File: test/test_types.py [hunk 6]
```diff
@@ -263,16 +263,13 @@ def test_TPM2B_PUBLIC(self):
...
-        self.assertEqual(b, b"test")
+        pub = TPM2B_PUBLIC(publicArea=TPMT_PUBLIC(nameAlg=TPM2_ALG.SHA256))
+        self.assertEqual(pub.publicArea.nameAlg, TPM2_ALG.SHA256)
 
         with self.assertRaises(
-            AttributeError, msg="TPM2B_DIGEST has no field by the name of badfield"
+            AttributeError, msg="TPM2B_PUBLIC has no field by the name of badfield"
         ):
-            TPM2B_DIGEST(badfield=1)
+            TPM2B_PUBLIC(badfield=1)
```
[the remark this answers is not in the document]
Comment (whooo) [reply test/test_types.py:272]: As I changed the init method for TPM2B_DIGEST I wanted to make sure that the this test was for the init method in TPM_OBJECT

File: tpm2_pytss/types.py (hunk 2 above)
Comment (whooo) [reply tpm2_pytss/types.py:1520]: Sounds good, will add it to the next pusj

File: tpm2_pytss/types.py (hunk 5 above)
Comment (whooo) [reply tpm2_pytss/types.py:941]: Using `_cdata.__setattr__(k, v)` in TPM_OBJECTs `__init__` breaks some functionality, then you need to pass cdata for everything

Comment (williamcroberts) [reply tpm2_pytss/types.py:941]: > Using `_cdata.__setattr__(k, v)` in TPM_OBJECTs `__init__` breaks some functionality, then you need to pass cdata for everything

Ahh Ok, according to https://lgtm.com/rules/10030085/, we want the base class fully initialized before calling to things that are overridden by the subclass. Perhaps we want TPM_OBJECT.__setattr__(self, k, v), so we specifically call the base class setattr here?

Comment (whooo) [reply tpm2_pytss/types.py:941]: I wonder something like `super().__init__(_cdata)` and then loop over kwargs and running `super().__setattr__(k, v)` in TPM2B_SIMPLE_OBJECT might work instead

Comment (whooo) [reply tpm2_pytss/types.py:941]: test_TPM2B_AUTH_bad_fields fails then

File: tpm2_pytss/types.py (hunk 4 above)
Comment (whooo) [reply tpm2_pytss/types.py:964]: I need to touch ESAPI.py as well when removing TPM2B_unpack, so will make another commit (under this PR) for that

File: tpm2_pytss/types.py (hunk 5 above)
Comment (whooo) [reply tpm2_pytss/types.py:942]: Found an ugly hack, lets see if LGTM will accept it

Comment (williamcroberts) [reply tpm2_pytss/types.py:942]: > Found an ugly hack, lets see if LGTM will accept it

Yeah I thought multiple init calls would trigger a warning, yep:
```
1 for Multiple calls to `__init__` during object initialization
1 for Unused import
1 for `__init__` method calls overridden method
```

On the plus side we got a different result :-p

Comment (williamcroberts) [reply tpm2_pytss/types.py:942]: > test_TPM2B_AUTH_bad_fields fails then

Is that the only test that fails? Maybe some logic change to the base class? It must be calling setattr on the object over cdata which will work. Perhaps in the constructor we limit the values that can be setattr'd to valid fields in the cdata?

Comment (whooo) [reply tpm2_pytss/types.py:941]: yeah, exactly, the field check is in the init method, so perhaps a `_setfield` method in TPM_OBJECT that only handles cdata fields and use that in its init method and  `__setattr__`

Comment (whooo) [reply tpm2_pytss/types.py:941]: yeah, the fields check is in `__init__`,

File: tpm2_pytss/types.py [hunk 7]
```diff
@@ -920,55 +920,52 @@ def Unmarshal(cls, buf):
...
-
-        # if their is nothing to update OR the caller passed both the
-        # buffer AND the size, their is nothing to do.
-        if len(kwargs) == 0 or len(kwargs) == 2:
-            return
-
-        key = [*kwargs][0]
-        value = kwargs[key]
-        if len(kwargs) == 0:
-            return
```
Comment (williamcroberts) [inline tpm2_pytss/types.py:941]: I'm a big fan of programming via the delete key.

**Review by williamcroberts — CHANGES_REQUESTED** [review]
LGTM I can handle the conflicts, or do you? We can fixup any other nits later.

**Review by williamcroberts — APPROVED** [review]

**Comment by whooo** [comment]
> LGTM I can handle the conflicts, or do you? We can fixup any other nits later.

I''ll take care of that, need to fix the TPM2B_unpack import anyway, but that will happen tomorrow (CET)

**Comment by williamcroberts** [comment]
> > LGTM I can handle the conflicts, or do you? We can fixup any other nits later.
> 
> I''ll take care of that, need to fix the TPM2B_unpack import anyway, but that will happen tomorrow (CET)

Okay, ill freeze master and wait for that, this way ill rebase on you.

**Comment by whooo** [comment]
LGTM-bot is still not happy about `__init__` :-/

Other then that everything is fixed, perhaps the warning is acceptable?

**Comment by williamcroberts** [comment]
> LGTM-bot is still not happy about `__init__` :-/
> 
> Other then that everything is fixed, perhaps the warning is acceptable?

Yeah, we'll get it later.
